CIP-XXXX/README.md

```diff
@@ -0,0 +1,270 @@
+Type: Standards
+Created: 2022-11-29
+License: <CC-BY-4.0>
+---
+
+## Abstract
+
+Currently, the hash of a script is simply the hash of its [serialisation](
+https://github.com/input-output-hk/plutus/blob/a645d1ee0dd5efcd7a7da24678461e07396ad26e/plutus-ledger-api/src/PlutusLedgerApi/Common/SerialisedScript.hs#L88).
+This CIP proposes changing this such that the hash of a script (term)
+is a function of its immediate children's hashes, forming a Merkle Tree from the AST.
+This allows one to shallowly verify a script's hash, and is useful on Cardano,
+because it allows scripts to **check that a script hash is an instantiation of a parameterised script**.
+
+In addition, a `blake2b_224` built-in function must be added.
+
+This is inspired by [BIP-144](https://github.com/bitcoin/bips/blob/master/bip-0114.mediawiki),
+but the motivations are very different.
+
+## Motivation
+
+Given some core logic expressible as a script, it is common to have parameters in
+the form of constants, e.g. fees, references to other scripts, magical numbers.
+
+These parameters can either be put in a datum somewhere, or can be put into the
+script itself, either by inlining them, or applying the unapplied script to the constants.
+
+On-chain it is currently hard to check that one script is an applied form
+of another script. In cases where that is necessary, datums are instead used.
+
+By Merkelising the hashing, we make this possible,
+which unlocks checking that a script is an application of another script to some parameter.
+
+Example reasons to apply the parameters to the script:
+- Staking validators currently don't support datums, and all staking validators
+  share a single rewards account. Allowing checking applied parameters
+  makes staking validators much more powerful. (More about this below)
+- Constants can be included in reference script, leading to less CPU and memory used,
+  since they don't have to be parsed from the adjacent datum (somewhat cheap)
+  or the script context (very expensive).
+- A script address + datum can't fit in an address,
+  if you want that you also need this (or need to change what an address is).
+
+## Specification
+
+The hash of a script will be derived directly from the AST, rather than its serialisation.
+Currently, its formed by hashing the serialisation prefixed with a byte that represents its version,  e.g. 0x02 for Plutus V2.
+
+The hash of a script becomes the hash of the prefix version annotation prepended to the hash of the term.
+
+[`Term`](https://github.com/input-output-hk/plutus/blob/a645d1ee0dd5efcd7a7da24678461e07396ad26e/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Type.hs#L69)
+currently has 8 constructors. On-chain, annotations are always the unit type,
+and are hence ignored for this algorithm. Each case/constructor is generally handled by
+hashing the concatenation of a prefix (single byte corresponding to the
+constructor index) along with the hashes of the arguments passed to the constructor.
```

Comment:
This is slightly different to what @kwxm wrote here (https://github.com/input-output-hk/plutus/blob/master/doc/notes/plutus-core/Merklisation/Merklisation-notes.md#modified-merklisation-technique), which I think also included the serialized versions of the nodes in the value that gets hashed. Not sure if that's important, Kenneth do you remember?